File: MyLeetCodeAnswers/C++/TwoSum.cpp
/*
Given an array of integers, find two numbers such that they add up to a specific target number.

The function twoSum should return indices of the two numbers such that they add up to the target, where index1 must be less than index2. Please note that your returned answers (both index1 and index2) are not zero-based.

You may assume that each input would have exactly one solution.

Input: numbers={2, 7, 11, 15}, target=9
Output: index1=1, index2=2
注意数组可能无序！
*/

class Solution {
public:
    vector<int> twoSum(vector<int> &numbers, int target) {
        unordered_map<int, int> map;
        vector<int> result;
        for (int i = 0; i < numbers.size(); ++i) {
            const int rest = target - numbers[i];
            if (map.find(rest) != map.end()) {
                result.push_back(map[rest]+1);// 先放查找池中的数字！
                result.push_back(i+1);
                break;
            }
            // 如果找不到对应的，将其放入到查找池中！%>_<%
            map[numbers[i]] = i;
        }
        return result;
    }
};

File: MyLeetCodeAnswers/java/TwoSum.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

/**
 * Source   : https://leetcode.com/problems/two-sum/
 * Author   : Wang Meng
 * Date     : 2016-03-15
 * <p>
 * Given an array of integers, return indices of the two numbers such that they add up to a specific target.
 * <p>
 * You may assume that each input would have exactly one solution.
 * <p>
 * Example:
 * Given nums = [2, 7, 11, 15], target = 9,
 * <p>
 * Because nums[0] + nums[1] = 2 + 7 = 9,
 * return [0, 1].
 */
public class TwoSum {
    public static int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; ++i) {
            map.put(nums[i], i);
        }
        for (int i = 0; i < nums.length; ++i) {
            if (map.containsKey(target - nums[i]) && map.get(target - nums[i]) > i) {
                return new int[]{i, map.get(target - nums[i])};
            }
        }
        return null;
    }

    public static void main(String[] args) {
        int[] nums = {7, 11, 15, 2};
        System.out.println(Arrays.toString(TwoSum.twoSum(nums, 9)));
    }
}
